File: LaundryLoop.Api/Controllers/BookingsController.cs

using LaundryLoop.Api.Data;
using LaundryLoop.Api.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace LaundryLoop.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class BookingsController : ControllerBase
{
    private readonly BookingRepository _repo;
    private readonly ILogger<BookingsController> _logger;

    public BookingsController(BookingRepository repo, ILogger<BookingsController> logger)
    {
        _repo   = repo;
        _logger = logger;
    }

    // POST /api/bookings
    [HttpPost]
    public async Task<IActionResult> Create([FromBody] CreateBookingRequest req)
    {
        if (!ModelState.IsValid)
            return BadRequest(ModelState);

        var booking = new Booking
        {
            Reference        = GenerateReference(),
            ApartmentName    = req.ApartmentName.Trim(),
            ApartmentAddress = req.ApartmentAddress.Trim(),
            UnitNumber       = req.UnitNumber.Trim(),
            BuzzerCode       = req.BuzzerCode?.Trim(),
            Phone            = req.Phone.Trim(),
            Email            = req.Email?.Trim().ToLowerInvariant(),
            SlotId           = req.SlotId,
            SlotDay          = req.SlotDay,
            SlotPeriod       = req.SlotPeriod,
            PickupTime       = req.PickupTime,
            DeliveryTime     = req.DeliveryTime,
            LaundrySize      = req.LaundrySize,
            TotalCost        = req.TotalCost,
            Notes            = req.Notes?.Trim(),
            Status           = "Pending",
            CreatedAt        = DateTime.UtcNow,
            UpdatedAt        = DateTime.UtcNow,
        };

        await _repo.InsertAsync(booking);

        // Insert cart items
        foreach (var item in req.CartItems)
        {
            var cartItem = new CartItem
            {
                BookingId = booking.Id,
                ItemType = item.ItemType,
                ItemName = item.ItemName,
                Description = item.Description,
                Price = item.Price
            };
            await _repo.InsertCartItemAsync(cartItem);
        }

        _logger.LogInformation("Booking created: {Ref} — {Apt} unit {Unit}",
            booking.Reference, booking.ApartmentName, booking.UnitNumber);

        return CreatedAtAction(nameof(GetById), new { id = booking.Id }, MapToResponse(booking));
    }

    // GET /api/bookings/{id}
    [HttpGet("{id:int}")]
    public async Task<IActionResult> GetById(int id)
    {
        var booking = await _repo.GetByIdAsync(id);
        return booking is null ? NotFound() : Ok(MapToResponse(booking));
    }

    // GET /api/bookings?page=1&pageSize=50&status=Pending&sortBy=apartmentAddress&sortDir=asc
    [Authorize]
    [HttpGet]
    public async Task<IActionResult> GetAll(
        [FromQuery] int page       = 1,
        [FromQuery] int pageSize   = 50,
        [FromQuery] string? status  = null,
        [FromQuery] string? sortBy  = "createdAt",
        [FromQuery] string? sortDir = "desc")
    {
        var (total, items) = await _repo.GetAllAsync(page, pageSize, status, sortBy, sortDir);

        return Ok(new
        {
            total,
            page,
            pageSize,
            items = items.Select(MapToResponse),
        });
    }

    // PATCH /api/bookings/{id}/status
    // Body: { "status": "PickedUp" }
    [Authorize]
    [HttpPatch("{id:int}/status")]
    public async Task<IActionResult> UpdateStatus(int id, [FromBody] UpdateStatusRequest req)
    {
        var allowed = new[] { "Pending", "Confirmed", "PickedUp", "Delivered", "Cancelled" };
        if (!allowed.Contains(req.Status))
            return BadRequest(new { error = $"Invalid status. Allowed: {string.Join(", ", allowed)}" });

        var booking = await _repo.UpdateStatusAsync(id, req.Status);
        return booking is null ? NotFound() : Ok(MapToResponse(booking));
    }

    // DELETE /api/bookings/{id}
    [Authorize]
    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        var deleted = await _repo.DeleteAsync(id);
        return deleted ? NoContent() : NotFound();
    }

    // Helpers
    private static string GenerateReference()
        => $"LLP-{Random.Shared.Next(100_000, 999_999)}";

    private static BookingResponse MapToResponse(Booking b) => new()
    {
        Id               = b.Id,
        Reference        = b.Reference,
        Status           = b.Status,
        CreatedAt        = b.CreatedAt,
        ApartmentName    = b.ApartmentName,
        ApartmentAddress = b.ApartmentAddress,
        UnitNumber       = b.UnitNumber,
        BuzzerCode       = b.BuzzerCode,
        Phone            = b.Phone,
        Email            = b.Email,
        SlotDay          = b.SlotDay,
        SlotPeriod       = b.SlotPeriod,
        PickupTime       = b.PickupTime,
        DeliveryTime     = b.DeliveryTime,
        LaundrySize     = b.LaundrySize,
        TotalCost        = b.TotalCost,
        Notes            = b.Notes,
    };


    // GET /api/bookings/{id}/cart-items
    [HttpGet("{id:int}/cart-items")]
    public async Task<IActionResult> GetCartItems(int id)
    {
        var items = await _repo.GetCartItemsAsync(id);
        return Ok(items);
    }

    // ─────────────────────────────────────────
    // GET /api/config
    // ─────────────────────────────────────────
    [HttpGet("/api/config")]
    public IActionResult GetConfig([FromServices] IConfiguration config)
    {
        var baseUrl = config["BaseUrl"] ?? "http://localhost:5000";
        return Ok(new { baseUrl });
    }
}

public record UpdateStatusRequest(string Status);

File: LaundryPickupDeliveryPHP/src/Controllers/BookingsController.php
<?php

declare(strict_types=1);

namespace LaundryLoop\Controllers;

use LaundryLoop\Config\Settings;
use LaundryLoop\Models\Booking;
use LaundryLoop\Models\CartItem;
use LaundryLoop\Repository\BookingRepository;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;
use Psr\Log\LoggerInterface;

class BookingsController
{
    public function __construct(
        private readonly BookingRepository $repo,
        private readonly LoggerInterface   $logger
    ) {}

    // ─── POST /api/bookings ────────────────────────────────────────────────────

    public function create(Request $request, Response $response): Response
    {
        $body = $request->getParsedBody();

        if (!is_array($body)) {
            return $this->json($response, ['error' => 'Invalid JSON body'], 400);
        }

        // Validate required fields — mirrors [Required] annotations on CreateBookingRequest
        $required = [
            'apartmentName', 'apartmentAddress', 'unitNumber',
            'phone', 'slotId', 'slotDay', 'slotPeriod',
            'pickupTime', 'deliveryTime', 'laundrySize', 'totalCost',
        ];

        $errors = [];
        foreach ($required as $field) {
            if (empty($body[$field]) && $body[$field] !== '0') {
                $errors[$field] = "$field is required";
            }
        }

        if (!empty($body['email']) && !filter_var($body['email'], FILTER_VALIDATE_EMAIL)) {
            $errors['email'] = 'Invalid email address';
        }

        if (!empty($errors)) {
            return $this->json($response, ['errors' => $errors], 400);
        }

        $now = gmdate('Y-m-d H:i:s');

        $booking                  = new Booking();
        $booking->reference        = $this->generateReference();
        $booking->apartmentName    = trim($body['apartmentName']);
        $booking->apartmentAddress = trim($body['apartmentAddress']);
        $booking->unitNumber       = trim($body['unitNumber']);
        $booking->buzzerCode       = isset($body['buzzerCode']) ? trim($body['buzzerCode']) : null;
        $booking->phone            = trim($body['phone']);
        $booking->email            = isset($body['email']) ? strtolower(trim($body['email'])) : null;
        $booking->slotId           = trim($body['slotId']);
        $booking->slotDay          = trim($body['slotDay']);
        $booking->slotPeriod       = trim($body['slotPeriod']);
        $booking->pickupTime       = trim($body['pickupTime']);
        $booking->deliveryTime     = trim($body['deliveryTime']);
        $booking->laundrySize      = trim($body['laundrySize']);
        $booking->totalCost        = (float) $body['totalCost'];
        $booking->notes            = isset($body['notes']) ? trim($body['notes']) : null;
        $booking->status           = 'Pending';
        $booking->createdAt        = $now;
        $booking->updatedAt        = $now;

        $this->repo->insert($booking);

        // Insert cart items
        $cartItems = $body['cartItems'] ?? [];
        foreach ($cartItems as $itemData) {
            $item              = new CartItem();
            $item->bookingId   = $booking->id;
            $item->itemType    = $itemData['itemType']    ?? '';
            $item->itemName    = $itemData['itemName']    ?? '';
            $item->description = $itemData['description'] ?? null;
            $item->price       = (float) ($itemData['price'] ?? 0);
            $this->repo->insertCartItem($item);
        }

        $this->logger->info('Booking created: {ref} — {apt} unit {unit}', [
            'ref'  => $booking->reference,
            'apt'  => $booking->apartmentName,
            'unit' => $booking->unitNumber,
        ]);

        // 201 Created with Location header — mirrors CreatedAtAction()
        $responseData = $booking->toResponse();
        return $this->json(
            $response->withHeader('Location', "/api/bookings/{$booking->id}"),
            $responseData,
            201
        );
    }

    // ─── GET /api/bookings/{id} ────────────────────────────────────────────────

    public function getById(Request $request, Response $response, array $args): Response
    {
        $booking = $this->repo->getById((int) $args['id']);

        if ($booking === null) {
            return $this->json($response, ['error' => 'Not found'], 404);
        }

        return $this->json($response, $booking->toResponse());
    }

    // ─── GET /api/bookings?page=1&pageSize=50&status=Pending&sortBy=createdAt&sortDir=desc
    // [Authorize] — protected by BasicAuthMiddleware applied at route level

    public function getAll(Request $request, Response $response): Response
    {
        $params   = $request->getQueryParams();
        $page     = max(1, (int) ($params['page']     ?? 1));
        $pageSize = max(1, min(200, (int) ($params['pageSize'] ?? 50)));
        $status   = $params['status']  ?? null;
        $sortBy   = $params['sortBy']  ?? 'createdAt';
        $sortDir  = $params['sortDir'] ?? 'desc';

        $result = $this->repo->getAll($page, $pageSize, $status, $sortBy, $sortDir);

        return $this->json($response, [
            'total'    => $result['total'],
            'page'     => $page,
            'pageSize' => $pageSize,
            'items'    => $result['items'],
        ]);
    }

    // ─── PATCH /api/bookings/{id}/status ──────────────────────────────────────
    // Body: { "status": "PickedUp" }
    // [Authorize] — protected by BasicAuthMiddleware applied at route level

    public function updateStatus(Request $request, Response $response, array $args): Response
    {
        $allowed = ['Pending', 'Confirmed', 'PickedUp', 'Delivered', 'Cancelled'];
        $body    = $request->getParsedBody();
        $status  = $body['status'] ?? '';

        if (!in_array($status, $allowed, true)) {
            return $this->json($response, [
                'error' => 'Invalid status. Allowed: ' . implode(', ', $allowed),
            ], 400);
        }

        $booking = $this->repo->updateStatus((int) $args['id'], $status);

        if ($booking === null) {
            return $this->json($response, ['error' => 'Not found'], 404);
        }

        return $this->json($response, $booking->toResponse());
    }

    // ─── DELETE /api/bookings/{id} ─────────────────────────────────────────────
    // [Authorize] — protected by BasicAuthMiddleware applied at route level

    public function delete(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->repo->delete((int) $args['id']);
        return $deleted
            ? $response->withStatus(204)
            : $this->json($response, ['error' => 'Not found'], 404);
    }

    // ─── GET /api/bookings/{id}/cart-items ────────────────────────────────────

    public function getCartItems(Request $request, Response $response, array $args): Response
    {
        $items = $this->repo->getCartItems((int) $args['id']);
        return $this->json($response, $items);
    }

    // ─── GET /api/config ──────────────────────────────────────────────────────

    public function getConfig(Request $request, Response $response): Response
    {
        return $this->json($response, [
            'baseUrl' => Settings::get('app.base_url'),
        ]);
    }

    // ─── Helpers ───────────────────────────────────────────────────────────────

    private function generateReference(): string
    {
        return 'LLP-' . random_int(100_000, 999_999);
    }

    private function json(Response $response, mixed $data, int $status = 200): Response
    {
        $response->getBody()->write(json_encode($data, JSON_UNESCAPED_UNICODE));
        return $response->withStatus($status)->withHeader('Content-Type', 'application/json');
    }
}
